BitSliceIndexing/bsi.go

```diff
@@ -497,3 +612,78 @@ func (b *BSI) NewBSIRetainSet(foundSet *roaring.Bitmap) *BSI {
 	wg.Wait()
 	return newBSI
 }
+
+func (b *BSI) Clone() *BSI {
+	return b.NewBSIRetainSet(b.eBM)
+}
+
+// Add
+//
+// In-place sum the contents of another BSI with this BSI, column wise.
+//
+func (b *BSI) Add(other *BSI) {
+
+	b.eBM.Or(other.eBM)
+	for i := 0; i < len(other.bA); i++ {
+		b.addDigit(other.bA[i], i)
+	}
+}
+
+func (b *BSI) addDigit(foundSet *roaring.Bitmap, i int) {
+
+	if i >= len(b.bA) {
+		b.bA = append(b.bA, roaring.NewBitmap())
+	}
+	carry := roaring.And(b.bA[i], foundSet)
+	b.bA[i].Xor(foundSet)
+	if carry.GetCardinality() > 0 {
+		if i+1 >= len(b.bA) {
+			b.bA = append(b.bA, roaring.NewBitmap())
+		}
+		b.addDigit(carry, i+1)
+	}
+}
+
+// TransposeWithCounts is a matrix transpose function that returns a BSI that has a columnID system defined by the values
+// contained within the input BSI.   Given that for BSIs, different columnIDs can have the same value.  TransposeWithCounts
+// is useful for situations where there is a one-to-many relationship between the vectored integer sets.  The resulting BSI
+// contains the number of times a particular value appeared in the input BSI as an integer count.
+//
+func (b *BSI) TransposeWithCounts(parallelism int, foundSet *roaring.Bitmap) *BSI {
+
+	return parallelExecutorBSIResults(parallelism, b, transposeWithCounts, foundSet, true)
+}
+
+func transposeWithCounts(input *BSI, batch []uint32, resultsChan chan *BSI, wg *sync.WaitGroup) {
+
+	defer wg.Done()
+
+	results := NewDefaultBSI()
+	if input.runOptimized {
+		results.RunOptimize()
+	}
+	for _, cID := range batch {
+		if value, ok := input.GetValue(uint64(cID)); ok {
+			if val, ok2 := results.GetValue(uint64(value)); !ok2 {
+				results.SetValue(uint64(value), 1)
+			} else {
+				val++
+				results.SetValue(uint64(value), val)
+			}
+		}
+	}
+	resultsChan <- results
+}
+
+// Increment
+//
+// In-place increment of values in a BSI.  Found set select columns for incrementing.
+// It assumes that the column value already exists.
+//
+func (b *BSI) Increment(foundSet *roaring.Bitmap) {
```

Comment:
The comment should be of the form "Increment does X"

Reply:
Done